### ISSUE TITLE
EgressGW: support CiliumEndpointSlices

### ISSUE THREAD

**Issue by qmonnet** [cilium/cilium#24833]
### What happened?

Tested on EKS.

When CiliumEndpointSlices are enabled, removing Egress Gateway policies does not behave as expected and rules are not created or removed in datapath's map.

I did the following:

1. Set up an EKS cluster, with both [CiliumEndpointSlices](https://docs.cilium.io/en/stable/network/kubernetes/ciliumendpointslice_beta/) and [Egress Gateway](https://docs.cilium.io/en/stable/network/egress-gateway/#selecting-the-destination) enabled.
2. Create a minimal Egress Gateway policy for one node, apply it.
3. At this point:
   - I can see the CRD loaded for Kubernetes with `kubectl get crd ciliumegressgatewaypolicies.cilium.io`.
   - But I can _not_ see any relevant rule in the output from `cilium bpf egress list` on the Agent pod.

Conversely, for deleting a policy that was added without CES:

1. Set up an EKS cluster, with Egress Gateway enabled but CiliumEndpointSlices disabled.
2. Create a minimal Egress Gateway policy for one node, apply it.
3. At this point:
   - I can see the CRD loaded for Kubernetes with `kubectl get crd ciliumegressgatewaypolicies.cilium.io`
   - I can also observe the relevant rules from `cilium bpf egress list` on the Agent pod.
4. Enable CiliumEndpointSlices.
5. At this point:
   - Kubernetes no longer sees the policy.
   - The rules are still in place in `cilium bpf egress list`.

Instead, I would expect rules to be added to the Egress Gateway map when the policy is created, and removed when it is deleted, independently of CiliumEndpointSlices being enabled.

Having the rules not created/deleted properly results in inconsistencies in the datapath's behaviour, and I've observed connectivity loss for some pods.

I note that Egress Gateway was not initially supported with CiliumEndpointSlices when the latter was introduced (see https://github.com/cilium/cilium/issues/17669), but that this was supposedly addressed in https://github.com/cilium/cilium/pull/19194 (cc @blzhao-0).

### Cilium Version

1.12.4

### Kernel Version

`5.4.238-148.346.amzn2.x86_64`

### Kubernetes Version

`v1.23.17-eks-ec5523e`

**Comment by pchaigno** [comment]
cc @Weil0ng @alan-kut

**Comment by alan-kut** [comment]
cc @dlapcevic @alexkats

**Comment by julianwiedmann** [comment]
(adjusting title / labels, as this is now a documented limitation that is enforced by the agent)

**Comment by jshr-w** [comment]
Hi, I'd like to pick up this issue, would someone be able to assign me please?

**Comment by marseel** [comment]
Some code pointers for you @jshr-w 

Here is where EGW currently is fetching CiliumEndpoint information: https://github.com/cilium/cilium/blob/43d65ed933e57fb079c9f960ab82b60fb36d46fd/pkg/egressgateway/manager.go#L328

This is handling only "regular" CEPs: https://github.com/cilium/cilium/blob/43d65ed933e57fb079c9f960ab82b60fb36d46fd/pkg/egressgateway/manager.go#L107

There is a similar resource pattern for CiliumEndpointSlices: https://github.com/cilium/cilium/blob/2311f3d7811c8950b7f5e1e2ac290f8216be0dc3/daemon/k8s/resources.go#L168

Here is how the handling of CES is done for other features:
- Events loop similar to regular CEP: https://github.com/cilium/cilium/blob/b0164d4da604fa6f39f4637cbab16269ea709b25/pkg/k8s/watchers/cilium_endpoint_slice.go#L51
- Here is the part that goes through these events: https://github.com/cilium/cilium/blob/bacf2e8718e603b6b74bd599711d73f3f7bfcddb/pkg/k8s/watchers/cilium_endpoint_slice_subscriber.go#L67
- which triggers update per each endpoint separately: https://github.com/cilium/cilium/blob/bacf2e8718e603b6b74bd599711d73f3f7bfcddb/pkg/k8s/watchers/cilium_endpoint_slice_subscriber.go#L172 - it has the same type that EGW is now expecting "*types.CiliumEndpoint"

Also, there is a check that crashed cilium when CES is enabled: https://github.com/cilium/cilium/blob/43d65ed933e57fb079c9f960ab82b60fb36d46fd/pkg/egressgateway/manager.go#L182

So these are some code-pointers how it is/works right now.

How I imagine it could work:
- EGW Instead of relying on its own informer https://github.com/cilium/cilium/blob/43d65ed933e57fb079c9f960ab82b60fb36d46fd/pkg/egressgateway/manager.go#L328
could instead just subscribe to events in:
- https://github.com/cilium/cilium/blob/9cd56a2fbeed048b6e2709f9f5a55fe18d3b0bfe/pkg/k8s/watchers/cilium_endpoint.go#L97
- and for delete events: https://github.com/cilium/cilium/blob/9cd56a2fbeed048b6e2709f9f5a55fe18d3b0bfe/pkg/k8s/watchers/cilium_endpoint.go#L207
In this way, it would be handling both regular CEP and CES transparently and even improving current scalability for EGW users that use CEP as agent's wouldn't open duplicated watchers.

@julianwiedmann do you have maybe some other pointers or better vision for supporting CES :) ?

**Comment by julianwiedmann** [comment]
> do you have maybe some other pointers or better vision for supporting CES :) ?

Not at all, I would also just start by looking how other consumers are handling it today :).

One thing to note is the caveat in https://github.com/cilium/cilium/pull/29124:
>The drawback of using UID is incompatibility with CiliumEndpointSlices, which is however not an issue for the EGW, as it is already incompatible. In the future the EGW will probably move away from relying on CiliumEndpoints altogether, which should remove this limitation.

**Comment by marseel** [comment]
> Not at all, I would also just start by looking how other consumers are handling it today :).

How everyone else is doing it is through subscribing to managers, like this:
- node: https://github.com/cilium/cilium/blob/2311f3d7811c8950b7f5e1e2ac290f8216be0dc3/pkg/node/manager/cell.go#L41
- endpoint: https://github.com/cilium/cilium/blob/2311f3d7811c8950b7f5e1e2ac290f8216be0dc3/pkg/endpointmanager/cell.go#L127

This would also include support for kvstore/clustermesh etc, so I think it's an order of magnitude harder - or at least it would require order of magnitude more testing.

**Comment by marseel** [comment]
I will unassign Shreya for now, as it's probably a bit too convoluted area of code right now.
